python/pyarrow/tests/test_cpp_internals.py

```diff
@@ -26,6 +31,10 @@ def inject_cpp_tests(ns):
     Inject C++ tests as Python functions into namespace `ns` (a dict).
     """
     for case in get_cpp_tests():
+        if np is None and ('numpy' in case.name or 'pandas' in case.name):
+            # Skip C++ tests that require pandas or numpy if numpy not present
+            continue
```

Comment:
As those are not test items at that point I wasn't sure how to add the markers here. I ended up adding a `pytest_collection_modifyitems` on `conftest.py` to add the corresponding markers. Let me know if this approach makes sense to you.

Reply:
I am not entirely sure this will work, but since you are essentially defining the test function here, I was wondering if something like this would work:

```
        def wrapper(request, case=case):
            if 'numpy' in case.name:
                request.node.add_marker(pytest.mark.numpy)
            case()
```

(and similar for pandas)

Reply:
It might be this does not work for _selecting_ tests, though, because the marker is only added when running the test (it does work for xfailing or skipping). 

So could either use that to skip the test directly (instead of through the `numpy` mark), or could also do something like the following (but that is more verbose):

```
if 'numpy' in case.name:

    @pytest.mark.numpy
    def wrapper():
        case()

else:

    def wrapper():
        case()
```

In the end this is also a decorator, so maybe the following also works (not entirely sure this works this way as those pytest decorators are a bit magical ..):

```
def wrapper():
    case()

if 'numpy' in case.name:
    wrapper = pytest.mark.numpy(wrapper)
```

Reply:
Thanks @jorisvandenbossche ! I've tested the proposed approach and validated it works. I've gone with that removing the `pytest_collection_modifyitems`

Reply:
Nice!